File: src/checks/estimates.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

export const estimateFromUsageCheck: Check = {
  name: 'Estimate From Usage',
  description: 'Estimate costs from historical usage',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);
    const customerId = ctx.testCustomerId || ctx.createdCustomerId;

    if (!customerId) {
      return {
        name: 'Estimate From Usage',
        success: false,
        duration: 0,
        message: 'No customer ID available',
        suggestion: 'Run Customer Create check first or set TEST_CUSTOMER_ID',
      };
    }

    try {
      const sdk = (client as unknown as {
        sdk: {
          estimateFromUsage?: (data: { customerId: string; startDate: string; endDate: string }) => Promise<{ estimatedCost: number; currency: string }>;
        };
      }).sdk;

      if (!sdk.estimateFromUsage) {
        const duration = performance.now() - start;
        return {
          name: 'Estimate From Usage',
          success: true,
          duration,
          message: 'Skipped (estimateFromUsage not available)',
          details: 'The estimateFromUsage method is not available in the SDK',
        };
      }

      const endDate = new Date().toISOString().split('T')[0];
      const startDate = new Date(Date.now() - 30 * 24 * 60 * 60 * 1000).toISOString().split('T')[0];

      const result = await sdk.estimateFromUsage({
        customerId,
        startDate,
        endDate,
      });
      const duration = performance.now() - start;

      return {
        name: 'Estimate From Usage',
        success: true,
        duration,
        message: `Estimated: ${result.estimatedCost} ${result.currency}`,
        details: `Period: ${startDate} to ${endDate}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Estimate From Usage',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Estimate From Usage',
        success: false,
        duration,
        message: err.message || 'Failed to estimate from usage',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const estimateFromHypotheticalCheck: Check = {
  name: 'Estimate Hypothetical',
  description: 'Estimate hypothetical costs',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const sdk = (client as unknown as {
        sdk: {
          estimateFromHypothetical?: (data: { items: Array<{ meter: string; quantity: number }> }) => Promise<{ estimatedCost: number; currency: string; breakdown: Array<{ meter: string; cost: number }> }>;
        };
      }).sdk;

      if (!sdk.estimateFromHypothetical) {
        const duration = performance.now() - start;
        return {
          name: 'Estimate Hypothetical',
          success: true,
          duration,
          message: 'Skipped (estimateFromHypothetical not available)',
          details: 'The estimateFromHypothetical method is not available in the SDK',
        };
      }

      const result = await sdk.estimateFromHypothetical({
        items: [
          { meter: 'tokens', quantity: 1000 },
          { meter: 'api_calls', quantity: 100 },
        ],
      });
      const duration = performance.now() - start;

      return {
        name: 'Estimate Hypothetical',
        success: true,
        duration,
        message: `Estimated: ${result.estimatedCost} ${result.currency}`,
        details: `Items: ${result.breakdown?.length || 2}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Estimate Hypothetical',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Estimate Hypothetical',
        success: false,
        duration,
        message: err.message || 'Failed to estimate hypothetical',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/estimates.py
"""Cost estimation checks."""
from datetime import datetime, timedelta
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _estimate_from_usage_check(ctx: CheckContext) -> CheckResult:
    """Estimate costs from historical usage."""
    customer_id = ctx.created_customer_id or ctx.test_customer_id
    if not customer_id:
        return CheckResult(
            name="estimate_from_usage",
            success=False,
            duration=0,
            message="No customer ID available",
            suggestion="Run customer_create check first or set TEST_CUSTOMER_ID"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'estimate_from_usage'):
            return CheckResult(
                name="estimate_from_usage",
                success=True,
                duration=0,
                message="Skipped (estimate_from_usage not available)",
                details="The estimate_from_usage method is not available in the SDK"
            )

        end_date = datetime.now().strftime('%Y-%m-%d')
        start_date = (datetime.now() - timedelta(days=30)).strftime('%Y-%m-%d')

        result = client.estimate_from_usage(
            customer_id=customer_id,
            start_date=start_date,
            end_date=end_date
        )

        estimated_cost = getattr(result, 'estimated_cost', getattr(result, 'estimatedCost', 'N/A'))
        currency = getattr(result, 'currency', 'USD')

        return CheckResult(
            name="estimate_from_usage",
            success=True,
            duration=0,
            message=f"Estimated: {estimated_cost} {currency}",
            details=f"Period: {start_date} to {end_date}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="estimate_from_usage",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="estimate_from_usage",
            success=False,
            duration=0,
            message=f"Failed to estimate from usage: {e}"
        )


estimate_from_usage_check = Check(
    name="estimate_from_usage",
    description="Estimate costs from historical usage",
    run=_estimate_from_usage_check
)


async def _estimate_from_hypothetical_check(ctx: CheckContext) -> CheckResult:
    """Estimate hypothetical costs."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'estimate_from_hypothetical'):
            return CheckResult(
                name="estimate_hypothetical",
                success=True,
                duration=0,
                message="Skipped (estimate_from_hypothetical not available)",
                details="The estimate_from_hypothetical method is not available in the SDK"
            )

        result = client.estimate_from_hypothetical(
            items=[
                {"meter": "tokens", "quantity": 1000},
                {"meter": "api_calls", "quantity": 100}
            ]
        )

        estimated_cost = getattr(result, 'estimated_cost', getattr(result, 'estimatedCost', 'N/A'))
        currency = getattr(result, 'currency', 'USD')
        breakdown = getattr(result, 'breakdown', [])

        return CheckResult(
            name="estimate_hypothetical",
            success=True,
            duration=0,
            message=f"Estimated: {estimated_cost} {currency}",
            details=f"Items: {len(breakdown) if breakdown else 2}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="estimate_hypothetical",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="estimate_hypothetical",
            success=False,
            duration=0,
            message=f"Failed to estimate hypothetical: {e}"
        )


estimate_from_hypothetical_check = Check(
    name="estimate_hypothetical",
    description="Estimate hypothetical costs",
    run=_estimate_from_hypothetical_check
)
